kani-compiler/src/kani_middle/coercion.rs

```diff
@@ -0,0 +1,188 @@
+use rustc_middle::ty::adjustment::CustomCoerceUnsized;
+use rustc_middle::ty::TypeAndMut;
+use rustc_middle::ty::{self, ParamEnv, TraitRef, Ty, TyCtxt};
+use rustc_span::symbol::Symbol;
+use tracing::trace;
+
+/// Extract the pair (`T`, `U`) for a unsized coercion where type `T` implements `Unsize<U>`.
+/// I.e., `U` is either a trait or a slice.
+/// For more details, please refer to:
+/// <https://doc.rust-lang.org/reference/type-coercions.html#unsized-coercions>
+///
+/// This is used to determine the vtable implementation that must be tracked by the fat pointer.
+///
+/// For example, if `&u8` is being converted to `&dyn Debug`, this method would return:
+/// `(u8, dyn Debug)`.
+///
+/// There are a few interesting cases (references / pointers are handled the same way):
+/// 1. Coercion between `&T` to `&U`.
+///    - This is the base case.
+///    - In this case, we extract the type that is pointed to.
+/// 2. Coercion between smart pointers like `Rc<T>` to `Rc<U>`.
+///    - Smart pointers implement the `CoerceUnsize` trait.
+///    - Use CustomCoerceUnsized information to traverse the smart pointer structure and find the
+///      underlying pointer.
+///    - Use base case to extract `T` and `U`.
+/// 3. Coercion between `&Wrapper<T>` to `&Wrapper<U>`.
+///    - Apply base case to extract the pair `(Wrapper<T>, Wrapper<U>)`.
+///    - Extract the tail element of the struct which are of type `T` and `U`, respectively.
+/// 4. Coercion between smart pointers of wrapper structs.
+///    - Apply the logic from item 2 then item 3.
+pub fn extract_unsize_casting<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    src_ty: Ty<'tcx>,
+    dst_ty: Ty<'tcx>,
+) -> (Ty<'tcx>, Ty<'tcx>) {
+    trace!(?dst_ty, ?src_ty, "find_trait_conversion");
+    let info = CoerceUnsizedIterator::new(tcx, src_ty, dst_ty).last().unwrap();
+    let src_base_ty = extract_pointee(info.src_ty)
+        .expect(&format!("Expected source to be a pointer. Found {:?} instead", info.src_ty));
+    let dst_base_ty = extract_pointee(info.dst_ty)
+        .expect(&format!("Expected destination to be a pointer. Found {:?} instead", info.dst_ty));
+    tcx.struct_lockstep_tails_erasing_lifetimes(src_base_ty, dst_base_ty, ParamEnv::reveal_all())
+}
+
+/// Iterates over the coercion path of a structure that implements `CoerceUnsized<T>` trait.
+/// The `CoerceUnsized<T>` trait indicates that this is a pointer or a wrapper for one, where
+/// unsizing can be performed on the pointee. More details:
+/// <https://doc.rust-lang.org/std/ops/trait.CoerceUnsized.html>
+///
+/// Given an unsized coercion between `impl CoerceUnsized<T>` to `impl CoerceUnsized<U>` where
+/// `T` is sized and `U` is unsized, this iterator will walk over the fields that lead to a
+/// pointer to `T`, which shall be converted from a thin pointer to a fat pointer.
+///
+/// This iterator will also include the name of the field that differs in the current iteration.
```

Comment:
where is this stored?

Reply:
It's not stored. It's generated as part of each iteration.